src/WebJobs.Script/Http/RetryProxyHandler.cs

```diff
@@ -30,20 +32,37 @@ public RetryProxyHandler(HttpMessageHandler innerHandler, ILogger logger)
 
         protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
         {
+            TaskCompletionSource<ScriptInvocationResult> resultSource = null;
+            if (request.Options.TryGetValue(ScriptConstants.HttpProxyScriptInvocationContext, out ScriptInvocationContext scriptInvocationContext))
+            {
+                resultSource = scriptInvocationContext.ResultSource;
+            }
+
             var currentDelay = InitialDelay;
             for (int attemptCount = 1; attemptCount <= MaxRetries; attemptCount++)
             {
                 try
                 {
+                    if (resultSource is not null && (resultSource.Task.IsFaulted || resultSource.Task.IsCanceled))
+                    {
+                        ExceptionDispatchInfo.Capture(resultSource.Task.Exception).Throw();
+                    }
+
                     return await base.SendAsync(request, cancellationToken);
                 }
                 catch (TaskCanceledException) when (cancellationToken.IsCancellationRequested)
                 {
                     _logger.LogDebug("Request was canceled. Stopping retries.");
                     throw new OperationCanceledException(cancellationToken);
                 }
-                catch (HttpRequestException) when (attemptCount < MaxRetries)
+                catch (HttpRequestException exception) when (attemptCount < MaxRetries)
                 {
+                    if (resultSource is not null && (resultSource.Task.IsFaulted || resultSource.Task.IsCanceled))
+                    {
+                        _logger.LogWarning("HTTP requests will not be retried. The associated function invocation has failed.");
+                        ExceptionDispatchInfo.Capture(exception).Throw();
```

Comment:
```suggestion
                        throw;
```

A simple `throw;` statement is a shorthand to do exactly this from within a `catch` block. It will rethrow the caught exception without resetting the stack trace.